cpp/src/arrow/extension/fixed_shape_tensor_test.cc

```diff
@@ -0,0 +1,188 @@
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/extension/fixed_shape_tensor.h"
+
+#include "arrow/testing/matchers.h"
+
+#include "arrow/array/array_nested.h"
+#include "arrow/array/array_primitive.h"
+#include "arrow/io/memory.h"
+#include "arrow/ipc/reader.h"
+#include "arrow/ipc/writer.h"
+#include "arrow/record_batch.h"
+#include "arrow/tensor.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/util/key_value_metadata.h"
+
+namespace arrow {
+
+class TestExtensionType : public ::testing::Test {};
+
+auto RoundtripBatch = [](const std::shared_ptr<RecordBatch>& batch,
+                         std::shared_ptr<RecordBatch>* out) {
+  ASSERT_OK_AND_ASSIGN(auto out_stream, io::BufferOutputStream::Create());
+  ASSERT_OK(ipc::WriteRecordBatchStream({batch}, ipc::IpcWriteOptions::Defaults(),
+                                        out_stream.get()));
+
+  ASSERT_OK_AND_ASSIGN(auto complete_ipc_stream, out_stream->Finish());
+
+  io::BufferReader reader(complete_ipc_stream);
+  std::shared_ptr<RecordBatchReader> batch_reader;
+  ASSERT_OK_AND_ASSIGN(batch_reader, ipc::RecordBatchStreamReader::Open(&reader));
+  ASSERT_OK(batch_reader->ReadNext(out));
+};
+
+TEST_F(TestExtensionType, FixedShapeTensorType) {
+  using FixedShapeTensorType = extension::FixedShapeTensorType;
+  using arrow::extension::fixed_shape_tensor;
+
+  std::vector<int64_t> shape = {3, 3, 4};
+  std::vector<int64_t> cell_shape = {3, 4};
+  auto value_type = int64();
+  std::shared_ptr<DataType> cell_type = fixed_size_list(value_type, 12);
+
+  std::vector<std::string> dim_names = {"x", "y"};
+  std::vector<int64_t> strides = {96, 32, 8};
+  std::vector<int64_t> column_major_strides = {8, 24, 72};
+  std::vector<int64_t> neither_major_strides = {96, 8, 32};
+  std::vector<int64_t> cell_strides = {32, 8};
+  std::vector<int64_t> values = {0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11,
+                                 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23,
+                                 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35};
+  std::vector<int64_t> values_partial = {0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11,
+                                         12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23};
+  std::vector<int64_t> shape_partial = {2, 3, 4};
+  std::string serialized = R"({"shape":[3,4],"dim_names":["x","y"]})";
+
+  ASSERT_OK_AND_ASSIGN(auto tensor,
+                       Tensor::Make(value_type, Buffer::Wrap(values), shape));
+  ASSERT_OK_AND_ASSIGN(
+      auto tensor_partial,
+      Tensor::Make(value_type, Buffer::Wrap(values_partial), shape_partial));
+
+  std::shared_ptr<ExtensionType> ext_type =
+      fixed_shape_tensor(value_type, cell_shape, {}, dim_names);
+  auto exact_ext_type = internal::checked_pointer_cast<FixedShapeTensorType>(ext_type);
+  ASSERT_OK_AND_ASSIGN(auto ds,
+                       ext_type->Deserialize(ext_type->storage_type(), serialized));
```

Comment:
Can you also add a test where passing a wrong type / wrong metadata to Deserialize?

Reply:
Added.